### PR TITLE
osd: refactor heartbeat health check

### PR DESCRIPTION
**Pull request by xiexingguo**
The idea was originally introduced two years ago in https://github.com/ceph/ceph/pull/9095.
We found it worked fair well if we intended to use a very short heartbeat interval(e.g., 1s) to detect
host or osd failure quickly and perform failover as fast as possible.

I am tired of backporting this from time to time. Since it has been widely verified in our products, I am taking another try here to (hopefully) get it into upstream to end my pain...

### CODE REVIEW

**Comment by xiexingguo** [comment]
@liewegas Can you take a look?

File: src/osd/OSD.h [hunk 1]
```diff
@@ -1508,20 +1508,25 @@ class OSD : public Dispatcher,
     utime_t last_rx_front;  ///< last time we got a ping reply on the front side
     utime_t last_rx_back;   ///< last time we got a ping reply on the back side
     epoch_t epoch;      ///< most recent epoch we wanted this peer
+    /// number of connections we send and receive heartbeat pings/replies
+    const static int HEARTBEAT_MAX_CONN = 2;
+    /// history of inflight pings, arranging by timestamp we sent
```
Comment (liewegas) [inline src/osd/OSD.h]: add a comment here like "send time -> deadline -> remaining replies"?

File: src/osd/OSD.cc [hunk 2]
```diff
@@ -4505,47 +4505,79 @@ void OSD::handle_osd_ping(MOSDPing *m)
...
+                     << " and older pending ping(s)"
+                     << dendl;
+
+            map<utime_t, pair<utime_t, int>> still_pending;
+            ++acked;
+            // only keep pings sent after this one.
+            // Note: acked may point to the end of ping_history and
+            // hence still_pending may stay empty after insertion.
+            still_pending.insert(acked, i->second.ping_history.end());
+            i->second.ping_history.swap(still_pending);
```
Comment (liewegas) [inline src/osd/OSD.cc]: wouldn't it be more efficient to just i->second.ping_history.erase(acked) (and drop the ++acked above)?  (or, if you want to erase all earlier acks, erase(i->second.ping_history.begin(), ++acked)?

**Comment by liewegas** [comment]
Otherwise this looks good!

**Comment by xiexingguo** [comment]
@liewegas Thanks!

**Pull request merged**